### PR TITLE
Update `find-stale-pods` logic

### PR DESCRIPTION
**Pull request by jbrower95**
- When computing the 5% deviation, we should be comparing against the balances of _all_ validators associated with the eigenpod (not just slashed ones).

### CODE REVIEW

File: cli/core/findStalePods.go [hunk 1]
```diff
@@ -204,30 +209,50 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
+		})
+
+		allValidatorBalancesSummed := utils.Reduce(allEigenpodsForValidator, func(accum phase0.Gwei, validator ValidatorWithIndex) phase0.Gwei {
+			return accum + allValidatorBalances[validator.Index]
+		}, phase0.Gwei(0))
+
+		balance := phase0.Gwei(cache.PodOwnerShares[eigenpod].NativeETH) + allValidatorBalancesSummed
+		allBalances[eigenpod] = phase0.Gwei(balance)
+		return allBalances
+	}, map[string]phase0.Gwei{})
```
Comment (jbrower95) [inline cli/core/findStalePods.go:235]: can i sum all the validator balances, plus the native eth?
or do I also need to filter the validators by which ones are active?

File: cli/core/findStalePods.go [hunk 2]
```diff
@@ -133,15 +136,15 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
+	allValidatorsWithIndices := utils.Map(_allValidators, func(validator *phase0.Validator, index uint64) ValidatorWithIndex {
 		return ValidatorWithIndex{
 			Validator: validator,
 			Index:     index,
 		}
 	})
 
 	// TODO(pectra): this logic changes after the pectra upgrade.
-	allSlashedValidators := aFilter(allValidatorsWithIndices, func(v ValidatorWithIndex) bool {
+	allSlashedValidators := utils.Filter(allValidatorsWithIndices, func(v ValidatorWithIndex) bool {
```
Comment (gpsanant) [inline cli/core/findStalePods.go:147]: we can do this before `allValidatorsWithIndices`

File: cli/core/findStalePods.go [hunk 3]
```diff
@@ -186,48 +191,71 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
 	}
 
-	var unhealthyEigenpods map[string]bool = make(map[string]bool)
-	for _, validator := range allActiveSlashedValidatorsBelongingToEigenpods {
-		balance := allValidatorBalances[validator.Index]
-		pod := validatorToPod[validator.Index]
-		executionBalance := cache.PodOwnerShares[pod].Shares
-		if executionBalance == 0 {
-			continue
+	totalAssetsGweiByEigenpod := utils.Reduce(keys(slashedEigenpods), func(allBalances map[string]phase0.Gwei, eigenpod string) map[string]phase0.Gwei {
```
Comment (jbrower95) [inline cli/core/findStalePods.go:220]: @gpsanant @wadealexc check this fn out

File: cli/core/findStalePods.go [hunk 4]
```diff
@@ -186,48 +191,71 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
-	for _, validator := range allActiveSlashedValidatorsBelongingToEigenpods {
+	slashedEigenpods := utils.Reduce(allActiveSlashedValidatorsBelongingToEigenpods, func(pods map[string][]ValidatorWithIndex, validator ValidatorWithIndex) map[string][]ValidatorWithIndex {
 		podAddress := executionWithdrawalAddress(validator.Validator.WithdrawalCredentials)
 		if podAddress != nil {
-			slashedEigenpods[*podAddress] = append(slashedEigenpods[*podAddress], validator)
+			if pods[*podAddress] == nil {
+				pods[*podAddress] = []ValidatorWithIndex{}
+			}
+			pods[*podAddress] = append(pods[*podAddress], validator)
 			validatorToPod[validator.Index] = *podAddress
```
Comment (gpsanant) [inline cli/core/findStalePods.go:210]: this is unused?

File: cli/core/findStalePods.go [hunk 5]
```diff
@@ -186,48 +191,71 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
-		executionBalance := cache.PodOwnerShares[pod].Shares
-		if executionBalance == 0 {
-			continue
+	totalAssetsGweiByEigenpod := utils.Reduce(keys(slashedEigenpods), func(allBalances map[string]phase0.Gwei, eigenpod string) map[string]phase0.Gwei {
+		// total assets of an eigenpod are determined as;
+		//	SUM(
+		//		- native ETH in the pod
+		//		- any active validators and their associated balances
+		// 	)
+		allEigenpodsForValidator := utils.Filter(allValidatorsWithIndices, func(v ValidatorWithIndex) bool {
```
Comment (gpsanant) [inline cli/core/findStalePods.go:226]: allValidatorsForEigenPod

File: cli/core/findStalePods.go [hunk 6]
```diff
@@ -25,8 +27,20 @@ type Cache struct {
...
+	out := make([]A, len(coll))
+	for key, _ := range coll {
+		out[len(out)] = key
+	}
+	return out
+}
+
 type PodOwnerShare struct {
 	Shares     uint64
+	NativeETH  phase0.Gwei
```
Comment (gpsanant) [inline cli/core/findStalePods.go:43]: call this ExecutionLayerBalance

File: cli/core/findStalePods.go [hunk 7]
```diff
@@ -85,10 +100,16 @@ func isEigenpod(eth *ethclient.Client, chainId uint64, eigenpodAddress string) (
...
+	balance, err := eth.BalanceAt(context.Background(), common.HexToAddress(eigenpodAddress), nil)
+	if err != nil {
+		return false, fmt.Errorf("balance check failed: %s", err.Error())
+	}
+
 	// Simulate fetching from contracts
 	// Implement contract fetching logic here
 	cache.PodOwnerShares[eigenpodAddress] = PodOwnerShare{
 		Shares:     podOwnerShares.Uint64(),
+		NativeETH:  phase0.Gwei(balance.Uint64()),
```
Comment (gpsanant) [inline cli/core/findStalePods.go:112]: need to divide by 1e9, balanceAt gives wei

File: cli/core/findStalePods.go [hunk 8]
```diff
@@ -186,48 +191,71 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
+			withdrawal := executionWithdrawalAddress(v.Validator.WithdrawalCredentials)
+			return withdrawal != nil && strings.EqualFold(*withdrawal, eigenpod)
+		})
+
+		allValidatorBalancesSummed := utils.Reduce(allEigenpodsForValidator, func(accum phase0.Gwei, validator ValidatorWithIndex) phase0.Gwei {
+			return accum + allValidatorBalances[validator.Index]
+		}, phase0.Gwei(0))
+
+		balance := phase0.Gwei(cache.PodOwnerShares[eigenpod].NativeETH) + allValidatorBalancesSummed
+		allBalances[eigenpod] = phase0.Gwei(balance)
```
Comment (gpsanant) [inline cli/core/findStalePods.go:236]: make this one line

File: cli/core/findStalePods.go [hunk 9]
```diff
@@ -25,8 +27,20 @@ type Cache struct {
 	PodOwnerShares map[string]PodOwnerShare
```
Comment (gpsanant) [inline cli/core/findStalePods.go:27]: This cache only lasts for the duration of a CLI command, right?

File: cli/core/findStalePods.go (hunk 7 above)
Comment (jbrower95) [reply cli/core/findStalePods.go:112]: phenomenal catch

File: cli/core/findStalePods.go (hunk 9 above)
Comment (jbrower95) [reply cli/core/findStalePods.go:32]: yep

File: cli/core/findStalePods.go [hunk 10]
```diff
@@ -85,11 +105,17 @@ func isEigenpod(eth *ethclient.Client, chainId uint64, eigenpodAddress string) (
...
+		return false, fmt.Errorf("balance check failed: %s", err.Error())
+	}
+
 	// Simulate fetching from contracts
 	// Implement contract fetching logic here
 	cache.PodOwnerShares[eigenpodAddress] = PodOwnerShare{
-		Shares:     podOwnerShares.Uint64(),
-		IsEigenpod: true,
+		Shares:                podOwnerShares.Uint64(),
+		ExecutionLayerBalance: weiToGwei(balance.Uint64()),
```
Comment (wadealexc) [inline cli/core/findStalePods.go:117]: Note: shares are in wei here, and you're converting the balance to gwei

File: cli/core/findStalePods.go [hunk 11]
```diff
@@ -186,48 +194,69 @@ func FindStaleEigenpods(ctx context.Context, eth *ethclient.Client, nodeUrl stri
...
+		log.Printf("%d EigenPods were slashed\n", len(slashedEigenpods))
+	}
+
+	unhealthyEigenpods := utils.Filter(keys(slashedEigenpods), func(eigenpod string) bool {
+		balance, ok := totalAssetsGweiByEigenpod[eigenpod]
+		if !ok {
+			return false
 		}
+		executionBalance := cache.PodOwnerShares[eigenpod].Shares
 		if balance <= phase0.Gwei(float64(executionBalance)*ACCEPTABLE_BALANCE_DEVIATION) {
```
Comment (wadealexc) [inline cli/core/findStalePods.go:251]: Here you are comparing gwei (`balance`) to wei (`executionBalance`)

**Review by wadealexc — APPROVED** [review]
See comments above -- once addressed, this LGTM

File: cli/core/findStalePods.go (hunk 10 above)
Comment (jbrower95) [reply cli/core/findStalePods.go]: updated everything to be explicitly wei or gwei

**Pull request merged**